File: .github/templates/template-processing.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-
# @AUTHOR: Chun-Jie Liu
# @CONTACT: chunjie.sam.liu.at.gmail.com
# @DATE: 2024-12-12 10:00:00
# @DESCRIPTION: Template for Python processing script
# @VERSION: v0.0.1

# Standard library imports
import os
import subprocess
from concurrent.futures import ProcessPoolExecutor
from pathlib import Path
from typing import Annotated, Optional

# Third-party imports
import polars as pl
import typer
from rich import print
from rich.logging import RichHandler
from rich.progress import (
    BarColumn,
    Progress,
    SpinnerColumn,
    TextColumn,
    TimeElapsedColumn,
)

# Constants and configuration
BASEDIR = Path("/liulab/chunjie/data/scMOCHA")
DATADIR = BASEDIR / "data"
OUTDIR = BASEDIR / "output"
MAX_WORKERS = 20

CONFIG = {
    "param1": "value1",
    "param2": "value2",
}

# Logging setup
import logging

logging.basicConfig(
    level=logging.INFO,
    format="%(message)s",
    datefmt="[%X]",
    handlers=[RichHandler(rich_tracebacks=True)],
)
logger = logging.getLogger("scMOCHA")


# Class definitions
class DataProcessor:
    """Process scMOCHA data with configurable parameters."""

    def __init__(self, config: dict):
        """Initialize processor with configuration."""
        self.config = config
        logger.info("Initialized DataProcessor")

    def process(self, data: pl.DataFrame) -> pl.DataFrame:
        """
        Process input data.

        Args:
            data: Input polars DataFrame

        Returns:
            Processed polars DataFrame
        """
        # Processing logic
        processed = data.filter(pl.col("value").is_not_null()).with_columns(
            (pl.col("value") * 2).alias("processed_value")
        )
        return processed


# Function definitions
def process_file(filepath: Path) -> Optional[pl.DataFrame]:
    """
    Process a single file.

    Args:
        filepath: Path to input file

    Returns:
        Processed DataFrame or None if processing fails
    """
    try:
        # Read data
        data = pl.read_csv(filepath)

        # Process
        processor = DataProcessor(CONFIG)
        result = processor.process(data)

        logger.info(f"Processed {filepath.name}: {result.shape[0]} rows")
        return result

    except Exception as e:
        logger.error(f"Failed to process {filepath}: {e}")
        return None


def process_sample(row: dict) -> dict:
    """
    Process a single sample (for parallel execution).

    Args:
        row: Dictionary containing sample metadata

    Returns:
        Dictionary with processing results
    """
    gseid = row["gseid"]
    srrid = row["srrid"]

    try:
        # Construct paths
        input_path = DATADIR / gseid / srrid / "input.csv"
        output_path = DATADIR / gseid / srrid / "output.parquet"

        # Check if input exists
        if not input_path.exists():
            logger.warning(f"Input file not found: {input_path}")
            return {
                "srrid": srrid,
                "status": "skipped",
                "reason": "missing_input",
            }

        # Process data
        data = pl.read_csv(input_path)
        processed = data.filter(pl.col("value").is_not_null())

        # Save result
        output_path.parent.mkdir(parents=True, exist_ok=True)
        processed.write_parquet(output_path)

        logger.info(f"Processed {srrid}: {processed.shape[0]} rows")
        return {"srrid": srrid, "status": "success", "nrow": processed.shape[0]}

    except Exception as e:
        logger.error(f"Failed to process {srrid}: {e}")
        return {"srrid": srrid, "status": "failed", "error": str(e)}


def process_batch(
    metadata_file: Path, max_workers: int = MAX_WORKERS
) -> pl.DataFrame:
    """
    Process batch of samples in parallel.

    Args:
        metadata_file: Path to metadata CSV file
        max_workers: Number of parallel workers

    Returns:
        DataFrame with processing results
    """
    # Read metadata
    metadata = pl.read_csv(metadata_file)
    logger.info(f"Loaded {metadata.shape[0]} samples from metadata")

    # Process in parallel with progress bar
    results = []
    with Progress(
        SpinnerColumn(),
        TextColumn("[progress.description]{task.description}"),
        BarColumn(),
        TextColumn("[progress.percentage]{task.percentage:>3.0f}%"),
        TimeElapsedColumn(),
    ) as progress:
        task = progress.add_task(
            "[cyan]Processing samples...", total=metadata.shape[0]
        )

        with ProcessPoolExecutor(max_workers=max_workers) as executor:
            futures = [
                executor.submit(process_sample, row)
                for row in metadata.iter_rows(named=True)
            ]

            for future in futures:
                result = future.result()
                results.append(result)
                progress.update(task, advance=1)

    # Convert results to DataFrame
    results_df = pl.DataFrame(results)

    # Log summary
    success_count = results_df.filter(pl.col("status") == "success").shape[0]
    failed_count = results_df.filter(pl.col("status") == "failed").shape[0]
    skipped_count = results_df.filter(pl.col("status") == "skipped").shape[0]

    logger.info("Processing summary:")
    logger.info(f"  Success: {success_count}")
    logger.info(f"  Failed: {failed_count}")
    logger.info(f"  Skipped: {skipped_count}")

    return results_df


# Main CLI function
def main(
    input_file: Annotated[str, typer.Argument(help="Input file path")],
    output_file: Annotated[
        str, typer.Option("--output", "-o", help="Output file path")
    ] = "output.parquet",
    gseid: Annotated[
        Optional[str], typer.Option("--gseid", "-g", help="GSE accession ID")
    ] = None,
    batch_mode: Annotated[
        bool,
        typer.Option("--batch", "-b", help="Process batch from metadata file"),
    ] = False,
    max_workers: Annotated[
        int, typer.Option("--workers", "-w", help="Number of parallel workers")
    ] = MAX_WORKERS,
    verbose: Annotated[
        bool, typer.Option("--verbose", "-v", help="Enable verbose logging")
    ] = False,
):
    """
    Process scMOCHA data files.

    Examples:
        # Process single file
        python script.py input.csv -o output.parquet

        # Process batch in parallel
        python script.py metadata.csv --batch --workers 10

        # Verbose mode
        python script.py input.csv -v
    """
    # Set logging level
    if verbose:
        logger.setLevel(logging.DEBUG)
        logger.debug("Verbose mode enabled")

    # Convert to Path
    input_path = Path(input_file)
    output_path = Path(output_file)

    # Validate input file exists
    if not input_path.exists():
        logger.error(f"Input file not found: {input_path}")
        raise typer.Exit(code=1)

    # Process data
    logger.info(f"Processing {input_path}...")

    if batch_mode:
        # Batch processing mode
        results = process_batch(input_path, max_workers=max_workers)

        # Save results
        output_path.parent.mkdir(parents=True, exist_ok=True)
        results.write_parquet(output_path)
        logger.info(f"Saved batch results to {output_path}")

    else:
        # Single file processing mode
        result = process_file(input_path)

        if result is None:
            logger.error("Processing failed")
            raise typer.Exit(code=1)

        # Save result
        output_path.parent.mkdir(parents=True, exist_ok=True)
        result.write_parquet(output_path)
        logger.info(f"Saved output to {output_path}")

    logger.info("✓ Processing complete!")


if __name__ == "__main__":
    typer.run(main)

File: .github/templates/template-processing.R
#!/usr/bin/env Rscript
# Metainfo ----------------------------------------------------------------
# @AUTHOR: Chun-Jie Liu
# @CONTACT: chunjie.sam.liu.at.gmail.com
# @DATE: 2024-12-12 10:00:00
# @DESCRIPTION: Template for preprocessing/processing script
# @VERSION: v0.0.1

# Library -----------------------------------------------------------------
load_pkg(jutils)
load_pkg(
  data.table,
  dplyr,
  tidyr,
  glue,
  fs,
  parallel,
  GetoptLong,
  logger,
  qs,
  DBI,
  duckdb
)

# args --------------------------------------------------------------------
# s: string, i: integer, f: float, !: boolean
# @: array, %: hash
GetoptLong.options(help_style = "two-column")
VERSION = "v0.0.1"

verbose = FALSE
gseid = "GSE123456"
basedir = "/liulab/chunjie/data/scMOCHA"
ncores = 20

GetoptLong(
  "gseid=s",
  "GSE accession ID",
  "basedir=s",
  "base directory path",
  "ncores=i",
  "number of cores for parallel processing",
  "verbose!",
  "print detailed messages"
)

# header ------------------------------------------------------------------
log_threshold(TRACE)
log_layout(layout_glue_colors)

# log_info("Starting processing for {gseid}...")

# function ----------------------------------------------------------------

# Processing function for parallel execution
fn_process_sample <- function(.srrid, .gseid, .basedir) {
  # Sample-level processing logic

  # Construct paths
  input_path <- fs::path(.basedir, "data", .gseid, .srrid, "input.qs")
  output_path <- fs::path(.basedir, "data", .gseid, .srrid, "output.qs")

  # Check if input exists
  if (!fs::file_exists(input_path)) {
    # log_warn("Input file not found for {.srrid}")
    return(NULL)
  }

  # Process data
  result <- tryCatch(
    {
      data <- import(input_path)

      # Apply transformations
      processed <- data |>
        dplyr::filter(!is.na(value)) |>
        dplyr::mutate(
          processed_value = value * 2
        )

      # Save result
      export(processed, output_path)

      # log_success("Processed {.srrid}: {nrow(processed)} rows")
      return(list(srrid = .srrid, status = "success", nrow = nrow(processed)))
    },
    error = function(e) {
      # log_error("Failed to process {.srrid}: {e$message}")
      return(list(srrid = .srrid, status = "failed", error = e$message))
    }
  )

  return(result)
}

# load data ---------------------------------------------------------------

# Set paths
datadir <- fs::path(basedir, "data", gseid)
metafile <- fs::path(datadir, "metadata.qs")

# Load metadata
if (!fs::file_exists(metafile)) {
  log_error("Metadata file not found: {metafile}")
  stop("Missing required metadata file")
}

metadata <- import(metafile)
# log_info("Loaded metadata: {nrow(metadata)} samples")

# body --------------------------------------------------------------------

# Get list of samples to process
sample_list <- metadata$srrid

# Parallel processing
# log_info("Processing {length(sample_list)} samples using {ncores} cores...")

results <- parallel::mclapply(
  X = sample_list,
  FUN = function(.srrid) {
    fn_process_sample(.srrid, gseid, basedir)
  },
  mc.cores = ncores
)

# Summarize results
results_summary <- data.table::rbindlist(results, fill = TRUE)

# log_info("Processing summary:")
# log_info("  Success: {sum(results_summary$status == 'success', na.rm = TRUE)}")
# log_info("  Failed: {sum(results_summary$status == 'failed', na.rm = TRUE)}")

# save --------------------------------------------------------------------

# Save processing summary
summary_file <- fs::path(datadir, "processing_summary.qs")
export(results_summary, summary_file)

# log_success("Processing complete! Summary saved to {summary_file}")
